embassy-usb-driver/src/lib.rs

```diff
@@ -345,10 +345,46 @@ pub trait ControlPipe {
 
 /// IN Endpoint trait.
 pub trait EndpointIn: Endpoint {
-    /// Write a single packet of data to the endpoint.
+    /// Write data to the endpoint.
+    ///
+    /// If the data is larger than the endpoint's maximum packet size, it will be transmitted as a
+    /// sequence of packets, where all but the last packet is guaranteed to be the maximum packet
+    /// size.
+    ///
+    /// A short packet always signifies the end of a USB transfer.  If you have a large transfer
```

Comment:
I think this is a bit awkward. I understand the goal is to support "streaming" transfers, but it ends up being awkward for both the streaming and non-streaming case:

- For non-streaming, caller still has to check if the data size was accidentally a multiple of 64, and then do a zero-length write in that case.
- For streaming, caller has to ensure all chunks except the last are multiples of 64. Chances are whatever's generating the streamed data is not generating it in clean chunks of 64, so the caller has to write logic to buffer the last "half-full" 64byte packet.

solutions

- Not support streaming at all. We have the high-level `write_transfer` which writes one entire transfer, and the low-level `write_packet` which writes one packet.
- Support streaming with an `is_last: bool` param. 

Maybe not supporting streaming wouldn't be such a big of a deal?

Reply:
This is just how the USB spec is defined.  This PR does not change the behavior here.

Even with the current write() API, users need to manually call `write()` with a 0-length buffer if they want to signal an early end of a transfer and they happen to have just sent a full packet.

Note that in many cases this doesn't really matter, because the endpoints have indicated the transfer length some other way (e.g., in the wLength field for control transfers, or in the HID report descriptor for HID transfers).  If the other side knows the transfer length to expect the explicit 0-length packet to indicate the end of the transfer isn't needed.  An explicit short packet is only needed to signify the end of the transfer when the other side isn't really expecting this or otherwise doesn't know the transfer length.

> Support streaming with an is_last: bool param.

I would be happy to add an extra parameter that asks the device to send an extra zero-length packet after the end of the transfer, if the transfer is a multiple of the max packet size.

I wouldn't call it `is_last` though, since this isn't always needed at the end of a transfer.  Maybe `send_zlp` or `indicate_early_end`?

Reply:
>This is just how the USB spec is defined. This PR does not change the behavior here.

The spec defines (assuming 64b max packet size):
- A "packet" is a unit containing 0..=64 bytes
- A "transfer" is a sequence of 64byte packets, followed by one <64byte packet (zero length if necessary).

My point is this function is neither "write packet" nor "write transfer", it's "something in between". It does *some* of the chunking of transfers into packets for you, but it leaves *some* of the logic to the caller. It's a leaky abstraction. 

It'd be cleaner if we had functions for "write packet" (does no chunking for you) and "write transfer" (does ALL chunking for you).

There's this edge case of when the transferred data length is exactly equal to the "expected transfer size" by the host, and is a multiple of 64, you don't have to send the ZLP. I'm not sure if it hurts to send it anyway in that case, though. It feels like we can always send it and that's it?

Reply:
I don't believe it is safe to simply always send the ZLP: I think this would incorrectly be treated as 2 separate transfers in some cases, with the ZLP being treated as a separate transfer of its own.

In particular, I don't think your definition of a transfer is exactly accurate: a transfer doesn't have to end in a short packet.  It can end in a maximum-size packet if the sender and receiver both agree on and understand the transfer length ahead of time.  The USB 2.0 spec just defines a transfer as "One or more bus transactions to move information between a software client and its function", and it basically seems to allow functions to use higher-level protocol knowledge to define transfer boundaries.  The lower level driver code therefore (unfortunately) cannot really know the start and end of a transfer without having higher level knowledge of the semantics of the data being sent.

For example, consider the HID specification.  The device defines the various possible reports it can send in its HID descriptor.  Therefore the host knows all possible reports that could be sent, and the maximum length of any report that the device might try to send.  If the longest report is 128 bytes, the host knows that if the device sends 2 64 byte packets this must be the end of the transfer, and that the next packet can't still be part of the same report transfer.

From section 8.4 of the [HID 1.11 spec](https://www.usb.org/document-library/device-class-definition-hid-111):

> * Only one report is allowed in a single USB transfer.
> * ...
> * All reports except the longest which exceed wMaxPacketSize for the endpoint
must terminate with a short packet. The longest report does not require a short
packet terminator.

Due to this protocol behavior, I think the higher level code is always going to have to pass in some indication of whether or not their transfer needs to be ended with a short packet or not.  The USB driver itself simply doesn't have enough knowledge of the data being sent to be able to decide this on its own.

All that said, I take your point about this API being somewhat in between "write one packet" and "write one transfer".  I do think this is still a useful API to provide to higher level code, as it allows them to send larger amounts of data at once while still splitting transfers into chunks if they desire.  Having a simple "write_transfer()" API to write a full transfer makes sense to me, but I think it still will require the caller to pass in a flag to specify whether the transfer needs to explicitly end in a short packet or not.  Note that at that point, the "write_transfer()" behavior is basically the same as this current API (apart from the addition of this extra flag).

Let me also send out a PR with my code for API definition for streaming read behavior as well, since I suspect you will probably also have feedback on it too.  It has similar slightly interesting behavior around whether caller requested to read data that is a multiple of the maximum packet size or not.  Thanks also for taking the time to review and provide feedback on these PRs.

Reply:
do you have a source confirming you MUST NOT send the ZLP? that HID quote says "does not require", not "must not send".

Reply:
Also, just to be clear: specifically in the case of HID it might be okay to always send an extra ZLP after a transfer with the longest report size.  Presumably the host would just treat this as an extra 0-length transfer and would just ignore it.  This doesn't really feel like a desirable thing to do though, and it seems like this isn't necessarily guaranteed to be a safe thing to do in general for other arbitrary protocol classes.

Reply:
> do you have a source confirming you MUST NOT send the ZLP? that HID quote says "does not require", not "must not send".

I think my earlier reply raced with yours, since I wrote it before I saw yours, but I think this is entirely up to the higher level protocol definition.  For HID I think it would be okay to send an extra ZLP, since the host would likely just ignore the extra empty transfer.  However, based on the USB spec I think it would be entirely valid for a vendor-defined protocol to have different (unexpected) behavior on receipt of an extra empty packet, since this could be treated as a separate transfer.

The underlying problem here is simply that knowing the USB transfer boundaries requires higher level knowledge of the data contents and semantics, and this can't be known at the embassy-usb-driver level.  If you receive 2 full packets followed by 1 short packet, it's entirely up to the higher level code to decide if this is 1 transfer, 2 transfers, or 3 transfers.  At the driver level we can identify that a short packet always means the end of a transfer, but we can't tell if a full packet is the end of a transfer or not.

I'll update this PR to add 2 separate methods to EndpointIn: one that just does a simple write of the data (what this current method does), and one that writes the data plus ensures it ends with a short packet (basically, call the simple write, followed by a ZLP if the length was a multiple of the MPS).  I think it's maybe slightly nicer to have 2 separate methods rather than 1 method with a parameter, but I'm happy to just have 1 method with a parameter if you prefer.
For the HID example, the HID class code would presumably want to use the simple write call when writing the longest report (which is always true if there is only a single HID report defined), and would want to use the ZLP version when writing shorter reports.